test/common/testlib.py

```diff
@@ -302,69 +391,51 @@ def switch_to_frame(self, name: str | None) -> None:
 
         The selector should select the <input type="file"/> element.
         Files is a list of absolute paths to files which should be uploaded.
         """
-        r = self.cdp.invoke("Runtime.evaluate", expression='document.querySelector(%s)' % jsquote(selector))
-        objectId = r["result"]["objectId"]
-        self.cdp.invoke("DOM.setFileInputFiles", files=files, objectId=objectId)
-
-    def raise_cdp_exception(
-        self, func: str, arg: str, details: Mapping[str, Any], trailer: str | None = None
-    ) -> Never:
-        # unwrap a typical error string
-        if details.get("exception", {}).get("type") == "string":
-            msg = details["exception"]["value"]
-        elif details.get("text", None):
-            msg = details.get("text", None)
-        else:
-            msg = str(details)
-        if trailer:
-            msg += "\n" + trailer
-        raise Error("%s(%s): %s" % (func, arg, msg))
+        element = self.bidi("script.evaluate", expression=f"ph_find({jsquote(selector)})",
+                            awaitPromise=False, target={"context": self.driver.context})["result"]
+        self.bidi("input.setFiles", context=self.driver.context, element=element, files=files)
 
     def inject_js(self, code: str) -> None:
         """Execute JS code that does not return anything
 
         :param code: a string containing JavaScript code
         :type code: str
         """
-        self.cdp.invoke("Runtime.evaluate", expression=code, trace=code,
-                        silent=False, awaitPromise=True, returnByValue=False, no_trace=True)
+        # this is redundant now; the difference used to be important with CDP
+        self.eval_js(code)
 
     def eval_js(self, code: str, no_trace: bool = False) -> Any:
         """Execute JS code that returns something
 
         :param code: a string containing JavaScript code
         :param no_trace: do not print information about unknown return values (default False)
         """
-        result = self.cdp.invoke("Runtime.evaluate", expression=code, trace=code,
-                                 silent=False, awaitPromise=True, returnByValue=True, no_trace=no_trace)
-        if "exceptionDetails" in result:
-            self.raise_cdp_exception("eval_js", code, result["exceptionDetails"])
-        _type = result.get("result", {}).get("type")
-        if _type == 'object' and result["result"].get("subtype", "") == "error":
-            raise Error(result["result"]["description"])
-        if _type == "undefined":
-            return None
-        if _type and "value" in result["result"]:
-            return result["result"]["value"]
-
-        if opts.trace:
-            print("eval_js(%s): cannot interpret return value %s" % (code, result))
-        return None
+        return self.bidi("script.evaluate", expression=code, quiet=no_trace,
+                         awaitPromise=True, target={"context": self.driver.context})["result"]
```

Comment:
I find the `awaitPromise=True` part a bit surprising... I would have expected to see that as a kwarg with the default as false...

Reply:
On the wire protocol you have to specify the value explicitly (both True and False), and I don't want to make the signature of `bidi` "magic" -- it can run any BiDi command, not just script.evaluate. I don't see how to simplify this in a sensible way, what do you mean?

Reply:
I meant that this function (on Browser, in testlib) should grow a `await_promise=` boolean kwarg that we forward through to `self.bidi()`

Reply:
Hmm -- we never expected CDP commands to return promises, the CDP driver [always waited on them](https://github.com/cockpit-project/cockpit/blob/main/test/common/chromium-cdp-driver.js#L319). Note that this doesn't mean "the command *must* return a promise", it just means "*if* the command returns a promise, wait for it", and that matches our test's expectations. If we ever need to handle a Promise on the Python side, we can add that (but I doubt it -- what are you going to do with it? these can't be serialized even)